pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go

```diff
@@ -78,155 +76,212 @@ func (r *AtlasDatabaseUserReconciler) Reconcile(context context.Context, req ctr
-		log.Infof("Not removing Atlas database user from Atlas as the '%s' annotation is set", customresource.ResourcePolicyAnnotation)
-	} else if err := r.deleteUserFromAtlas(dbUser, project, log); err != nil {
-		log.Error("Failed to remove database user from Atlas: %s", err)
+func (r *AtlasDatabaseUserReconciler) handleDeletion(
+	ctx context.Context,
+	dbUser *mdbv1.AtlasDatabaseUser,
+	project *mdbv1.AtlasProject,
+	atlasClient mongodbatlas.Client,
+	log *zap.SugaredLogger,
+) (bool, workflow.Result) {
+	if dbUser.GetDeletionTimestamp().IsZero() {
+		return false, workflow.OK()
 	}
 
-	// We ignore the error as it will be printed by the function
-	_ = connectionsecret.RemoveStaleSecretsByUserName(r.Client, project.ID(), dbUser.Spec.Username, *dbUser, log)
+	if customresource.HaveFinalizer(dbUser, customresource.FinalizerLabel) {
+		err := connectionsecret.RemoveStaleSecretsByUserName(r.Client, project.ID(), dbUser.Spec.Username, *dbUser, log)
+		if err != nil {
+			return true, workflow.Terminate(workflow.DatabaseUserConnectionSecretsNotDeleted, err.Error())
+		}
+	}
 
-	return nil
-}
+	if customresource.IsResourceProtected(dbUser, r.ObjectDeletionProtection) {
+		log.Info("Not removing Atlas database user from Atlas as per configuration")
 
-func (r AtlasDatabaseUserReconciler) deleteUserFromAtlas(dbUser *mdbv1.AtlasDatabaseUser, project *mdbv1.AtlasProject, log *zap.SugaredLogger) error {
-	connection, err := atlas.ReadConnection(log, r.Client, r.GlobalAPISecret, project.ConnectionSecretObjectKey())
+		err := customresource.ManageFinalizer(ctx, r.Client, dbUser, customresource.UnsetFinalizer)
+		if err != nil {
+			return true, workflow.Terminate(workflow.AtlasFinalizerNotRemoved, err.Error())
+		}
+
+		return true, workflow.OK()
+	}
+
+	_, err := atlasClient.DatabaseUsers.Delete(context.Background(), dbUser.Spec.DatabaseName, project.ID(), dbUser.Spec.Username)
 	if err != nil {
-		return err
+		var apiError *mongodbatlas.ErrorResponse
+		if errors.As(err, &apiError) && apiError.ErrorCode != atlas.UsernameNotFound {
+			return true, workflow.Terminate(workflow.DatabaseUserNotDeletedInAtlas, err.Error())
+		}
+
+		log.Info("Database user doesn't exist or is already deleted")
 	}
 
-	atlasClient, err := atlas.Client(r.AtlasDomain, connection, log)
+	err = customresource.ManageFinalizer(ctx, r.Client, dbUser, customresource.UnsetFinalizer)
 	if err != nil {
-		return fmt.Errorf("cannot build Atlas client: %w", err)
+		return true, workflow.Terminate(workflow.AtlasFinalizerNotRemoved, err.Error())
 	}
 
-	userName := dbUser.Spec.Username
+	return true, workflow.OK()
+}
+
+func (r *AtlasDatabaseUserReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("AtlasDatabaseUser").
+		For(&mdbv1.AtlasDatabaseUser{}, builder.WithPredicates(r.GlobalPredicates...)).
+		Watches(&source.Kind{Type: &corev1.Secret{}}, watch.NewSecretHandler(r.WatchedResources)).
+		Complete(r)
+}
```

Comment:
Are you sure removing works fine in this case? I mean, won't it trigger the reconcile with result like "can not reconcile, the resource has been modified/deleted`?

Reply:
Yes, All the flows has worked as expected